### ISSUE TITLE
PSR2.ControlStructures.SwitchDeclaration does not handle if branches with returns

### ISSUE THREAD

**Issue by Majkl578** [squizlabs/PHP_CodeSniffer#834]
Consider the following code as an example:

``` php
<?php

switch ($foo) {
    case 1:
        if ($bar === 1) {
            return 1;
        } elseif ($bar === 2) {
            return 2;
        } else {
            return 3;
        }
    case 2:
        return 4;
}
```

Although the `case 1` body always returns and thus is not fall-through, the error is generated: `There must be a comment when fall-through is intentional in a non-empty case body.`

**Comment by gsherwood** [comment]
I'm not sure how hard this will be to fix. I might be able to just look for ELSE statements directly inside the case and search for returns in there, but I would also have to recursively check other nested ELSE statements inside as well.

In the meantime, the most obvious suggestion would be to make the return explicit, which also adheres to the common no-else-return rule (or even the no-else and return-early rules) because the ELSE is technically unnecessary:

```php
switch ($foo) {
    case 1:
        if ($bar === 1) {
            return 1;
        } elseif ($bar === 2) {
            return 2;
        }

        return 3;
    case 2:
        return 4;
}
```

**Comment by Majkl578** [comment]
I understand this is not easy to fix, one has to check all branches recursively. Explicit return is of course a solution, and also more clean. :)

By the way, this problem also applies to `throw`.

**Comment by dereuromark** [comment]
@Majkl578 Or continue ;)
But yeah, one should probably avoid this kind of nesting, better to refactor into a sub-method here.

**Comment by gsherwood** [comment]
Fixed by PR #1363
